alpha/declcfg/load.go

```diff
@@ -111,20 +115,113 @@ func walkFiles(root fs.FS, fn func(root fs.FS, path string, err error) error) er
+	if root == nil {
+		return nil, fmt.Errorf("no declarative config filesystem provided")
+	}
+
+	concurrency := runtime.NumCPU()
+
+	var (
+		fcfg     = &DeclarativeConfig{}
+		pathChan = make(chan string, concurrency)
+		cfgChan  = make(chan *DeclarativeConfig, concurrency)
+	)
+
+	// Create an errgroup to manage goroutines. The context is closed when any
+	// goroutine returns an error. Goroutines should check the context
+	// to see if they should return early (in the case of another goroutine
+	// returning an error).
+	eg, ctx := errgroup.WithContext(context.Background())
+
+	// Walk the FS and send paths to a channel for parsing.
+	eg.Go(func() error {
+		return sendPaths(ctx, root, pathChan)
+	})
+
+	// Parse paths concurrently. The waitgroup ensures that all paths are parsed
+	// before the cfgChan is closed.
+	var wg sync.WaitGroup
+	for i := 0; i < concurrency; i++ {
+		wg.Add(1)
+		eg.Go(func() error {
+			defer wg.Done()
+			return parsePaths(ctx, root, pathChan, cfgChan)
+		})
+	}
+	eg.Go(func() error {
+		// We do not need to check for context cancellation here because
+		// all path parsers will return and decrement the waitgroup counter
+		// when the context is cancelled, thus unblocking this call to Wait.
+		wg.Wait()
+		close(cfgChan)
+		return nil
+	})
+
+	// Merge parsed configs into a single config.
+	eg.Go(func() error {
+		return mergeCfgs(ctx, cfgChan, fcfg)
+	})
+
+	// Wait for all goroutines to finish.
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	return fcfg, nil
+}
+
+func sendPaths(ctx context.Context, root fs.FS, pathChan chan<- string) error {
+	defer close(pathChan)
+	return walkFiles(root, func(_ fs.FS, path string, err error) error {
 		if err != nil {
 			return err
 		}
-		cfg.Packages = append(cfg.Packages, fcfg.Packages...)
-		cfg.Channels = append(cfg.Channels, fcfg.Channels...)
-		cfg.Bundles = append(cfg.Bundles, fcfg.Bundles...)
-		cfg.Others = append(cfg.Others, fcfg.Others...)
+		select {
+		case pathChan <- path:
+		case <-ctx.Done(): // don't block on sending to pathChan
+			return ctx.Err()
+		}
 		return nil
-	}); err != nil {
-		return nil, err
+	})
+}
+
+func parsePaths(ctx context.Context, root fs.FS, pathChan <-chan string, cfgChan chan<- *DeclarativeConfig) error {
+	for {
+		select {
+		case <-ctx.Done(): // don't block on receiving from pathChan
+			return ctx.Err()
+		case path, ok := <-pathChan:
+			if !ok {
+				return nil
+			}
+			cfg, err := LoadFile(root, path)
+			if err != nil {
+				return err
+			}
+			select {
+			case cfgChan <- cfg:
+			case <-ctx.Done(): // don't block on sending to cfgChan
+				return ctx.Err()
+			}
+		}
+	}
+}
+
+func mergeCfgs(ctx context.Context, cfgChan <-chan *DeclarativeConfig, fcfg *DeclarativeConfig) error {
+	for {
+		select {
+		case <-ctx.Done(): // don't block on receiving from cfgChan
+			return ctx.Err()
+		case cfg, ok := <-cfgChan:
+			if !ok {
+				return nil
+			}
+			fcfg.Packages = append(fcfg.Packages, cfg.Packages...)
```

Comment:
`go test -race` might false positive on this write if it's accessed elsewhere

Reply:
Standard `make unit` uses `-race`, and I haven't seen any issues, so hopefully all good.

I know you said false positive, but and I'm sure how the race detector works, but I don't see any race possibilities with `fcfg`.